File: omada b/askisi3/Plane.cpp
#include <iostream>
#include <cstdlib>
#include "ask3.h"

using namespace std;

int Plane::rand_PassCompartments(){
        static int i=(rand()%3)+2; /* Plane Passenger Compartments range [2,4] */
        return i;
}


Plane::Plane(string s,int c) :
        title(s)
        ,capacity(c)
        ,pcptr(new PassengerCompartment* [rand_PassCompartments()])
        ,cargoptr(new CargoBay)
        ,eqptr(new EquipmentCompartment* [3])
{
        for(int i=0; i<rand_PassCompartments(); i++)
                pcptr[i]=new PassengerCompartment;

        for(int i=0; i<3; i++)
                eqptr[i]=new EquipmentCompartment;

        cout<<endl<<endl<<"Plane: ";
        title.print();
        cout<< "Just created!"<<endl<<"~Capacity:"<<capacity<<endl<<endl;

}

Plane::Plane(int tid,string s,int c) :
        Object(tid)
        ,title(s)
        ,capacity(c)
        ,pcptr(new PassengerCompartment* [rand_PassCompartments()])
        ,cargoptr(NULL)
        ,eqptr(new EquipmentCompartment* [3])
{
        for(int i=0; i<rand_PassCompartments(); i++)
                pcptr[i]=NULL;

        for(int i=0; i<3; i++)
                eqptr[i]=NULL;

        cout<<endl<<endl<<"Plane: ";
        title.print();
        cout<< "Just created!"<<endl<<"~Capacity:"<<capacity<<endl<<endl;
}

Plane::~Plane(){
        cout<<"Plane to be destroyed!"<<endl;
        delete cargoptr;
        for(int i=0; i<3; i++)
                delete eqptr[i];
        delete[] eqptr;
        for(int i=0; i<rand_PassCompartments(); i++)
                delete pcptr[i];
        delete[] pcptr;
}


Object* Plane::clone() const {
/* Create a clone of "this"(Object) and return it */
/* I have to clone all the Plane Compartments of "this"  */
        Plane* p=new Plane(id,title.get_str(),capacity);

        for(int i=0; i<rand_PassCompartments(); i++) {
                p->pcptr[i]=(PassengerCompartment *)pcptr[i]->clone();
        }

        p->cargoptr=(CargoBay *)cargoptr->clone();

        for(int i=0; i<3; i++) {
                p->eqptr[i]=(EquipmentCompartment *)eqptr[i]->clone();
        }

        return p;

}


int Plane::equal(const Object& o) const {
/* Compare all data members */
/* Ι am using downcast */

        int flag=1;

        if(id==o.get_id()) {
                Plane *pPlane=(Plane *)&o; /* downcast */

                for(int i=0; i<rand_PassCompartments(); i++) {
                        if(pcptr[i]->equal(*(pPlane->get_pcptr()[i]))==0) {
                                flag=0;
                                break;
                        }
                }
                if(title.get_str()==pPlane->get_title().get_str()
                   && capacity==pPlane->get_capacity()
                   && cargoptr->equal(*(pPlane->get_cargoptr()))
                   && eqptr[0]->equal(*(pPlane->get_eqptr()[0]))
                   && eqptr[1]->equal(*(pPlane->get_eqptr()[1]))
                   && eqptr[2]->equal(*(pPlane->get_eqptr()[2]))
                   && flag ) {

                        return 1;
                }
        }
        return 0;
}


int Plane::identical(const Object* o) const {
/* Compare the addresses of two objects and the addresses of all Compartments */
/* Ι am using downcast */
        if (this==o) {
                int flag=1;

                Plane *pPlane=(Plane *)o;   /* downcast */

                for(int i=0; i<rand_PassCompartments(); i++) {
                        if(pcptr[i]->identical((pPlane->get_pcptr()[i]))==0) {
                                flag=0;
                                break;
                        }
                }
                if(cargoptr->identical((pPlane->get_cargoptr()))
                   && eqptr[0]->identical((pPlane->get_eqptr()[0]))
                   && eqptr[1]->identical((pPlane->get_eqptr()[1]))
                   && eqptr[2]->identical((pPlane->get_eqptr()[2]))
                   && flag ) {

                        return 1;
                }

        }
        return 0;
}


Mystring Plane::toString() const {
        /* Description for plane */
        Mystring s("Plane: id=");
        s=s.get_str().append(to_string(id));
        s=s.get_str().append(" Title:");
        s.concat(title);
        s=s.get_str().append(" Capacity:");
        s=s.get_str().append(to_string(capacity));
        /* For the compartments I call their own toStrings
           and then join them to the original (one by one) */
        for(int i=0; i<rand_PassCompartments(); i++)
                s.concat(pcptr[i]->toString());
        s.concat(cargoptr->toString());
        for(int i=0; i<3; i++)
                s.concat(eqptr[i]->toString());

        return s;
}

void Plane::process(SecurityEmployee& e){
/* Plane Compartments:Cargo Bay,Passenger Compartment and Equipment Compartment */
/* At these Plane Compartments provided Security Employee to works */

        cout<<endl<<"~Security Employee is in the plane!"<<endl<<endl;

        cargoptr->process(e);

        cargoptr->get_eqptr()->process(e); /* Cargo's Bay Equipment Compartment */

        for(int i=0; i<rand_PassCompartments(); i++) {
                pcptr[i]->process(e);
        }

        for(int i=0; i<3; i++) {
                eqptr[i]->process(e);
        }

}

void Plane::process(MaintenanceEmployee& e){
/* Plane Compartments:Cargo Bay and Equipment Compartment */
/* At these Plane Compartments provided Maintenance Employee to works */

        cout<<endl<<"~Maintenance Employee is in the plane!"<<endl<<endl;

        cargoptr->process(e);

        cargoptr->get_eqptr()->process(e); /* Cargo's Bay Equipment Compartment */

        for(int i=0; i<3; i++) {
                eqptr[i]->process(e);
        }

}

void Plane::process(CleaningEmployee& e){
/* Plane Compartments:Cargo Bay and Passenger Compartment */
/* At these Plane Compartments provided Cleaning Employee to works */

        cout<<endl<<"~CleaningEmployee Employee is in the plane!"<<endl<<endl;

        for(int i=0; i<rand_PassCompartments(); i++) {
                pcptr[i]->process(e);
        }

        cargoptr->process(e);


}

void Plane::ready_check(){
/* Check all Compartments if they have provided to them the suitable Employees */
/* Yes, then the plane is ready to take off */

        int flag=1;
        for(int i=0; i<rand_PassCompartments(); i++) {
                if(pcptr[i]->ready_check()==0) {
                        flag=0;
                        break;
                }
        }

        if(flag && cargoptr->ready_check() &&
           eqptr[0]->ready_check() && eqptr[1]->ready_check() && eqptr[2]->ready_check())
                cout<<endl<<endl<<"~~~~~~~~Plane ready to take off!~~~~~~~~"<<endl<<endl;
        else
                cout<<endl<<endl<<"~~~~~~~~Plane is NOT ready to take off!~~~~~~~~"<<endl<<endl;

}

File: omada b/askisiJava/Plane_package/Plane.java
package Plane_package;

import java.util.Random; /* to generate random numbers */
import Employee_package.*;
import Plane_Component_package.*;

public class Plane {

private int id;
private String title;
private int capacity;
private int Plane_Compartments;
private PassengerCompartment[] pc;
private CargoBay cb;
private EquipmentCompartment[] eq;

/* ----------------------------------- */


 public Plane(String s){
  Random rand = new Random();
  id=this.hashCode(); /* unique */
  title=s;
  capacity=rand.nextInt(101)+100; /* range [100,200] */
  Plane_Compartments=rand.nextInt(3)+2;  /* Plane Compartments range [2,4] */

  pc=new PassengerCompartment[Plane_Compartments];
  for (int i=0; i<Plane_Compartments; i++){
     pc[i]=new PassengerCompartment(1);
   }

  cb=new CargoBay();

  eq=new EquipmentCompartment[3];
  for (int i=0; i<3; i++){
     eq[i]=new EquipmentCompartment();
   }
  System.out.println("\n\nPlane: " + title +" just created!\n");
}


public void process(SecurityEmployee e){
/* Plane Compartments:Cargo Bay,Passenger Compartment and Equipment Compartment */
/* At these Plane Compartments provided Security Employee to works */

  System.out.println("\n~Security Employee is in the plane!\n");

  cb.process(e);

  cb.get_eq().process(e); /* Cargo's Bay Equipment Compartment */

  for (int i=0; i<Plane_Compartments; i++){
    pc[i].process(e);

  }

  for (int i=0; i<3; i++){
    eq[i].process(e);
  }

}
public void process(MaintenanceEmployee e){
/* Plane Compartments: Cargo Bay and Equipment Compartment */
/* At these Plane Compartments provided Maintenance Employee to works */

        System.out.println("\n~Maintenance Employee is in the plane!\n");

        cb.process(e);

        cb.get_eq().process(e); /* Cargo's Bay Equipment Compartment */

        for (int i=0; i<3; i++){
          eq[i].process(e);
        }

}
public void process(CleaningEmployee e){
/* Plane Compartments: Cargo Bay and Passenger Compartment */
/* At these Plane Compartments provided Cleaning Employee to works */
        System.out.println("\n~CleaningEmployee Employee is in the plane!\n");

        for (int i=0; i<Plane_Compartments; i++){
          pc[i].process(e);
        }

        cb.process(e);

}
public void ready_check(){
/* Check all Compartments if they have provided to them the suitable Employees */
/* Yes, then the plane is ready to take off */

        int flag=1;
        for (int i=0; i<Plane_Compartments; i++){
          if(pc[i].ready_check()==0) {
                  flag=0;
                  break;
          }
        }

        if(flag==1 && cb.ready_check()==1 &&
           eq[0].ready_check()==1 && eq[1].ready_check()==1 && eq[2].ready_check()==1)
           System.out.println("\n~~~~~~~~Plane ready to take off!~~~~~~~~\n");
        else
           System.out.println("\n~~~~~~~~Plane NOT ready to take off!~~~~~~~~\n");

}

@Override
public String toString() {
  /* Plane Description */
    String stemp=String.format("Plane: id=" + id +"\nTitle: "+ title +"\nCapacity: "+capacity);

    /* Description for Passengers Compartments */
    for (int i=0; i<Plane_Compartments; i++){
      stemp=stemp.concat("\n"+pc[i].toString());
    }

    /* Description for Cargo Bays */
    stemp=stemp.concat("\n"+cb.toString());

    /* Description for Equipment Compartments */
    for (int i=0; i<3; i++){
      stemp=stemp.concat("\n"+eq[i].toString());
    }

    return stemp;
}

//copy constructor
public Plane(Plane pl_old){

  id=pl_old.get_id();
  title=pl_old.get_title();
  capacity=pl_old.get_capacity();
  Plane_Compartments=pl_old.get_Plane_Compartments();

  pc=new PassengerCompartment[Plane_Compartments];
  for (int i=0; i<Plane_Compartments; i++){
     pc[i]=(PassengerCompartment) pl_old.get_pc()[i].clone();
   }

   cb=(CargoBay) pl_old.get_cb();

   eq=new EquipmentCompartment[3];
   for (int i=0; i<3; i++){
      eq[i]=(EquipmentCompartment) pl_old.get_eq()[i].clone();
    }
   System.out.println("\n\nPlane: " + title +" just created by copy!\n");

}

@Override
public boolean equals(Object obj){
  Plane temp=(Plane) obj;

  if(id==temp.get_id()){

    if(title==temp.get_title() &&
       capacity==temp.get_capacity() &&
       Plane_Compartments==temp.get_Plane_Compartments() &&
       cb.equals(temp.get_cb()) ){
         for(int i=0;i<Plane_Compartments;i++){
           if(pc[i].equals(temp.get_pc()[i])==false)
            return false;
         }
         for(int i=0;i<3;i++){
           if(eq[i].equals(temp.get_eq()[i])==false)
            return false;
         }

         return true;
       }
  }
  return false;
}

@Override
public Object clone(){
    return new Plane(this);
}

public String get_title(){
  return title;
}

public int get_capacity(){
  return capacity;
}

public int get_Plane_Compartments(){
  return Plane_Compartments;
}

public PassengerCompartment[] get_pc(){
  return pc;
}

public CargoBay get_cb(){
  return cb;
}

public int get_id(){
  return id;
}

public EquipmentCompartment[] get_eq(){
  return eq;
}

}
